pkg/execution/pauses/iterator.go

```diff
@@ -0,0 +1,244 @@
+	// is being used.
+	usingBuffer bool
+
+	// bufferIter is the buffered iterator.
+	bufferIter state.PauseIterator
+
+	// blockReader is the block reader to fetch metadata and blocks.
+	blockReader BlockReader
+
+	// unfetchedBlocks represents the set of blocks that haven't been fetched
+	// from the backing store yet.
+	unfetchedBlocks []ulid.ULID
+
+	// inflightBlocks represents blocks that are currently being fetched from
+	// the backing store.
+	inflightBlocks []ulid.ULID
+
+	// pauses represents the current pauses that have been fetched from downloaded
+	// blocks.
+	pauses []*state.Pause
+
+	err error
+
+	// l represents a lock held when mutating block slices or pauses.
+	l *sync.Mutex
+}
+
+// Count returns the count of the pause iteration at the time of querying.
+//
+// Due to concurrent processing, the total number of iterated fields may not match
+// this count;  the count is a snapshot in time.
+func (d *dualIter) Count() int {
+	return d.count
+}
+
+// Next advances the iterator, returning an erorr or context.Canceled if the iteration
+// is complete.
+//
+// Next should be called prior to any call to the iterator's Val method, after
+// the iterator has been created.
+//
+// The order of the iterator is unspecified.
+func (d *dualIter) Next(ctx context.Context) bool {
+	// Always attempt to fetch blocks if there's space.  This runs in the
+	// background.  We do this before calling bufferIter to ensure that we
+	// have older pauses loaded when the buffer iteration has finished.
+	//
+	// Iteration does NOT need to happen in-order.
+	d.fetchNextBlocks()
+
+	// Check on the buffer iterator and advance on this.
+	if d.bufferIter.Next(ctx) {
+		d.usingBuffer = true
+		return true
+	}
+
+	d.usingBuffer = false
+
+	// NOTE: We must release the lock as soon as possible such that the fetchBlock
+	// background thread can grab the lock to adjust pauses.
+	d.l.Lock()
+	quit := len(d.pauses) == 0 && len(d.inflightBlocks) == 0
+	d.l.Unlock()
+
+	if quit {
+		// We are done!  There are no pauses downloaded or inflight.
+		return false
+	}
+
+	// Complex case:  we may be waiting for in flight blocks to download, but only
+	// if we've handled all already downloaded pauses:
+	//
+	// eg:  0 blocks downloaded / 10 being downloaded.  spin until len(d.pauses) > 0.
+	d.l.Lock()
+	spin := len(d.inflightBlocks) > 0 && len(d.pauses) == 0
+	d.l.Unlock()
+
+	for spin {
+		// Wait 100ms for the block to download and try again.
+		time.Sleep(100 * time.Millisecond)
+
+		d.l.Lock()
+		if d.err != nil {
+			d.l.Unlock()
+			// Skip as we've errored.
+			return false
+		}
+
+		spin = len(d.inflightBlocks) > 0 && len(d.pauses) == 0
+		d.l.Unlock()
+	}
+
+	// Now we're good.
+	d.l.Lock()
+	defer d.l.Unlock()
+	if len(d.pauses) > 0 {
+		// Simple case:  pauses are downloaded, so yeah we have a next value.
+		return true
+	}
+
+	// If we're here and we've already processed all in-flight blocks, attempt
+	// to redownload blocks.
+	return d.Next(ctx)
```

Comment:
won't there be risk of deadlocks if this is a recursive call because of the `defer` a couple lines above?

Reply:
aha, fixed - thank you!